### PR TITLE
[Caffe2] Add ADD operator for IDEEP

### PR DESCRIPTION
**Pull request by yinghai**
It's basically a clone of `Sum`.

### CODE REVIEW

File: caffe2/ideep/operators/elementwise_sum_op.cc [hunk 1]
```diff
@@ -12,16 +12,23 @@ class IDEEPSumOp final : public IDEEPOperator {
...
     auto* Y = Output(OUTPUT);
 
     if (InputSize() == 1) {
       ideep::direct_copy::compute(X, *Y);
 
     } else {
       vector<itensor> inputs;
-      vector<float> scales (InputSize(), 1.0);
+      vector<float> scales(InputSize(), 1.0);
+      const auto dims = Input(0).get_dims();
```
Comment (bddppq) [inline caffe2/ideep/operators/elementwise_sum_op.cc]: nit: Input(0) -> X

File: caffe2/ideep/operators/elementwise_sum_op.cc [hunk 2]
```diff
@@ -12,16 +12,23 @@ class IDEEPSumOp final : public IDEEPOperator {
...
       for (int i = 0; i < InputSize(); ++i) {
+        if (Input(i).get_dims() != dims) {
+          CAFFE_ENFORCE_EQ(
+              dims,
+              Input(i).get_dims(),
+              "Broadcast is not yet supported with MKLDNN.");
```
Comment (bddppq) [inline caffe2/ideep/operators/elementwise_sum_op.cc]: ideep?

File: caffe2/ideep/operators/elementwise_sum_op.cc [hunk 3]
```diff
@@ -12,16 +12,23 @@ class IDEEPSumOp final : public IDEEPOperator {
...
+    const auto& X = Input(INPUT0);
     auto* Y = Output(OUTPUT);
 
     if (InputSize() == 1) {
       ideep::direct_copy::compute(X, *Y);
 
     } else {
       vector<itensor> inputs;
-      vector<float> scales (InputSize(), 1.0);
+      vector<float> scales(InputSize(), 1.0);
```
Comment (bddppq) [inline caffe2/ideep/operators/elementwise_sum_op.cc]: nit: const

File: caffe2/ideep/operators/elementwise_sum_op.cc [hunk 4]
```diff
@@ -12,16 +12,23 @@ class IDEEPSumOp final : public IDEEPOperator {
   virtual ~IDEEPSumOp() {}
 
   bool RunOnDevice() override {
-    const auto &X = Input(INPUT0);
+    const auto& X = Input(INPUT0);
     auto* Y = Output(OUTPUT);
 
     if (InputSize() == 1) {
       ideep::direct_copy::compute(X, *Y);
```
Comment (bddppq) [inline caffe2/ideep/operators/elementwise_sum_op.cc]: does it handle inplace?

**Review by vishar0 — APPROVED** [review]

File: caffe2/ideep/operators/elementwise_sum_op.cc (hunk 2 above)
Comment (yinghai) [reply caffe2/ideep/operators/elementwise_sum_op.cc]: This actually bonafide `MKLDNN`. :D

File: caffe2/ideep/operators/elementwise_sum_op.cc (hunk 4 above)
Comment (yinghai) [reply caffe2/ideep/operators/elementwise_sum_op.cc]: You mean what if `Y == X`? Good question. Let's ask Intel folks. How do we check Inplace? Just by their blob name?

Comment (viswanathgs) [reply caffe2/ideep/operators/elementwise_sum_op.cc]: You can check by comparing blob pointers such as `if (&X == Y)`. IDEEP does appear to support for certain ops, but it's work checking how sum is handled.

**Review by bddppq — APPROVED** [review]

**Comment by yinghai** [comment]
Merge to unblock. Meanwhile I filed an issue asking about the inplace behavior: https://github.com/intel/ideep/issues/10

**Pull request merged**